board/technologic/ts7100/parse_strap.c

```diff
@@ -0,0 +1,225 @@
+/*
+ * Copyright (C) 2019-2022 Technologic Systems dba embeddedTS
+ *
+ * SPDX-License-Identifier:     GPL-2.0+
+ */
+
+/*
+ * Knowledge of strapping is encapsulated in this file.
+ */
+
+#include <asm/arch/iomux.h>
+#include <asm/arch/imx-regs.h>
+#include <asm/arch/mx6-pins.h>
+#include <asm/arch/sys_proto.h>
+#include <asm/gpio.h>
+#include <common.h>
+#include "tsfpga.h"
+#include "parse_strap.h"
+
+#define	NAND_CE0_B	IMX_GPIO_NR(4, 13)	/* Bit 0 / IO opt bit 0 */
+#define	UART2_TX_DATA	IMX_GPIO_NR(1, 20)	/* Bit 1 / IO opt bit 1 */
+#define	UART5_TX_DATA	IMX_GPIO_NR(1, 30)	/* Bit 2 / IO opt bit 2 */
+#define	UART4_TX_DATA	IMX_GPIO_NR(1, 28)	/* Bit 3 / IO opt bit 3 */
+#define	UART3_TX_DATA	IMX_GPIO_NR(1, 24)	/* Bit 4 / IO model bit 0 */
+#define	LCD_DATA08	IMX_GPIO_NR(3, 13)	/* Bit 5 / IO model bit 1 */
+#define	NAND_CE1_B	IMX_GPIO_NR(4, 14)	/* Bit 6 / IO model bit 2 */
+#define	UART3_CTS_B	IMX_GPIO_NR(1, 26)	/* Bit 7 / IO model bit 3 */
+#define	JTAG_TDO_N15	IMX_GPIO_NR(1, 12)	/* Rev B on pad N15 */
+
+/*
+ * For Rev A: IO_MODEL is non-zero
+ * For Rev B: MX6UL_PAD_JTAG_TDO__GPIO1_IO12 (pad N15) reads as 1 instead of 0
+ *
+ * Five straps on the CPU board go into the FPGA:
+ *   RAM size: 
+ *     IO_B0 (C6)
+ *   CPU opts:
+ *     DONE_IO_B0 (A13)
+ *     IO_B0_SCL? (C8)
+ *     IO_B0 (C9)
+ *     IO_B0_SDA? (B8),
+ */
+
+const char *get_board_model(void)
+{
+  return "7100";
+}
+
+const char *get_board_name(void)
+{
+  uint8_t io_model;
+
+  io_model = read_io_board_model();
+  if (io_model == 0) {
+    return "TS-7100";
+  } else if (io_model == 1) {
+    return "TS-7100-Z";
+  }
+  return "TS-7100";
+}
+
+const char *get_cpu_board_version(void)
+{
+	uint8_t io_model;
+	uint8_t cpu_board_rev;
+        uint16_t raw_cpu_straps;
+        uint16_t raw_fpga_straps;
+	static char model_str[16] = {0};
+
+	io_model = read_io_board_model();
+        raw_cpu_straps =  read_raw_cpu_straps();
+        raw_fpga_straps =  read_raw_fpga_straps();
+
+        cpu_board_rev = ((raw_cpu_straps >> 7) & 0x02);    // Rev B 0x02 strap on the 6UL
+        cpu_board_rev |= ((raw_fpga_straps >> 12) & 0x01); // Rev A 0x01 strap on the FPGA
+
+        if (cpu_board_rev & 0x2) {
+          snprintf(model_str, sizeof(model_str), "B");
+          return model_str;
+        }
+        if (cpu_board_rev & 0x1) {
+          snprintf(model_str, sizeof(model_str), "A");
+          return model_str;
+        }
+        if (io_model == 0) {
+          snprintf(model_str, sizeof(model_str), "P2-%02x", cpu_board_rev);
+        } else if (io_model == 1) {
+          return "Am";
+        } else if (io_model == 3) {
+          snprintf(model_str, sizeof(model_str), "B-%02x", cpu_board_rev);
+        } else {
+          snprintf(model_str, sizeof(model_str), "UNKNOWN_%02x", cpu_board_rev);
+        }
+        return model_str;
```

Comment:
We talked about this one a little bit yesterday, but I'm going to suggest we do something like this:

```suggestion
const char get_cpu_board_version(void)
{
        uint8_t rev_b_strap;
        
	io_model = read_io_board_model();
        raw_cpu_straps =  read_raw_cpu_straps();
        rev_b_strap =  = ((raw_cpu_straps >> 7) & 0x02);

        if (rev_b_strap) {
                return 'B';
        } else {
                return 'A';
        }
}
```

This makes it a bit cheaper/simpler to do the rev comparisons when we need to instead of doing full string operations, eg:

```
if(get_cpu_board_version() == 'A') {
...
```

Reply:
Since putting up this code, I modified it to do basically what you're saying.

The assumption is we'll check bits from newest to oldest known hardware and return immediately when there is a "hit". Boards newer than rev B **should** return rev B.

I do check the "`rev_a_strap`" and fall-through to handle cases where something doesn't "add up". It should never be reached on hardware we release, but is a reality where speedy debug-ability has value. Meaning, a valid rev "A" should not be returned for something that is definitely wrong. If it is newer than B, we've already said it should return B, not A. If there is a mistake, we should make it easy to detect and track down.

< Lionel